### PR TITLE
Implement redis queue using python-rq

### PR DESCRIPTION
**Pull request by maethu**
**Goals with this PR:**
- [x] Add Redis queue for async indexing
- [x] Indexing should not make the site slower
- [x] Enable/Disable use of redis queue via control panel
- [x] Index binaries thru elastic search addon
- [x] Use multiple queues one for regular indexing and one for binaries.
- [x] Worker on separate cloud instances (For example a different docker container than the Plone instance)
- [x] Zero impact change on current installations. 
- [x] Run all current tests with redis as queue as well
- [x] Opt-in to use the Redis queue add any time. 
- [x] Provide docker config for a full setup (redis, worker, elasticsearch, maybe plone)
- [x] Detailed documentation for the new feature
- [x] Find good defaults for "low" and "normal" queue, retries, timeouts, etc. 

**Motivation:**
I will implement elastic search for a customer with roughly a 100 authors and hundreds to thousands of content manipulation a day. With the current solution collective.solr there is a performance impact especially when working with a large number of files (+100k). Copy/Paste/Move takes a long time. Also collective.solr has a ton of features I don't really need. This package has a much smaller code base and IMHO much simpler to use. The async indexing though, could be easily adapted to collective.solr as well. 

**Technical decisions so far - please challenge**
- Use Redis as queue: Mainly because the cloud provider I will use, provides Redis as DBAAS. Which means a less complex setup, less docker containers, etc. 
- [python-rq](https://python-rq.org/) as lib to control jobs creation and workers: This lib was implemented for Redis and much more lightweight than for example celery. I went thru the docs of both and did sample integrations, python-rq, at least for, had the easer and better api and I was fast to do the integration.
- Create jobs ind after commit hook: This is necessary, since longer commits may take too long and the rq worker starts working too early. Using the after commit hook, we can make sure the DB is synced and up to date.
- Extracting data using the plone.restapi: My goal ist have the worker run on a different instance/container. The worker container will have the same code base, so pickling/unpickling will work. 
- Environ variables: The necessary config for the Redis queue is all in env vars (Redis, DSN, user/pw, plone backend URL.
- Not using any existing integrations / packages: Well it's always fun to write new code 🙈  - There was no up to date package with a simple API, providing what I needed. At least I was not able to find one. 
- Binary data gets appended to the searchableText field via elasticsearch pipeline. The pipeline is configured as default pipeline on the index for plone
- Using cbor2 for data encoding/decoding. This avoids on average an 20% overhead (with base64) for every binary send to elasticsearch
- 2 queues: low and normal. The "normal" queue has a higher priority than the "low", rq worker can work first on the normal queue and then on the low one. 
- Redis connection gets handled thread safe as well, same as the elasticsearch connection

**Testing**
- I don't use the Redis fake connection, but an actual Redis. The queue though runs in "sync" mode https://python-rq.org/docs/#bypassing-workers 
- Using parameterized_class for current tests, in order to run them automatically with Redis/Queue as well. 

**Help needed**
I'm happy to do the work and my company is happy to invest the money needed to get this going.
But I need technical feedback and others hopefully testing the new feature in real life with real data.
- [x] ~How handle the redis connection (Plone instance / worker instance)~
- [ ] Best practices I'm missing.
- [x] ~Probably with elasticsearch addons (extracting the data)~

### CODE REVIEW

**Comment by maethu** [comment]
@ericof 
I hope it's ok to reach out to you before I finalize this PR. Maybe you have any input while I'm still implementing the final features and details. Really any input is really appreciated. 
 Any help going thru the process to have a final release at some point as well! 
Cheers,
Mathias

**Comment by maethu** [comment]
@ericof I can't satisfy black and flake8 analysis at the same time? Any hint how I can fix this?

**Comment by ericof** [comment]
@maethu 
```
portal_path_len = len(api.portal.get().getPhysicalPath())
obj_segements = obj.getPhysicalPath()
relative_path = "/".join(obj_segements[portal_path_len:])
```

**Comment by maethu** [comment]
@ericof Sorry it took a bit to finish this PR/Change from my end.
What do you think is the best way to proceed? Should I try to get some feedback as well thru community.plone.org ?

**Comment by ericof** [comment]
@maethu As this is a big PR, I will take some time to review it, ok?

**Comment by ericof** [comment]
@andreclimaco Could you test this branch against your installation?

**Comment by maethu** [comment]
> @maethu As this is a big PR, I will take some time to review it, ok?

@ericof Thank you very much!! I really appreciate it!! I'm happy about any feedback and how I can improve it!

**Comment by andreclimaco** [comment]
> @andreclimaco Could you test this branch against your installation?

Yes, I will install and test.

**Comment by andreclimaco** [comment]
> Could you test this branch against your installation?

@ericof, I tested the branch on my installation and didn't find any problems. Due to lack of time, I couldn't test the integration with redis.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -39,7 +57,7 @@ jobs:
 
       - name: Install package
         run: |
-          pip install -e ".[test]"
+          pip install -e ".[test, redis]"
```
Comment (wesleybl) [inline .github/workflows/tests.yml:60]: It might be good to run the tests with and without `redis`. Maybe have a `make build-with-redis`

**Comment by wesleybl** [comment]
I don't have time to look at the code, but I tested it with and without redis and it worked.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (maethu) [reply .github/workflows/tests.yml:60]: @wesleybl All tests are automatically executed with and without redis 😄  See https://github.com/collective/collective.elasticsearch/blob/mle-redis-rq/src/collective/elasticsearch/tests/test_processor.py#L15-L20

**Review by ericof — APPROVED** [review]

**Pull request merged**